### PR TITLE
Partial summation in bounce integrals

### PR DESCRIPTION
**Pull request by unalmis**
- [x] Applies (#1530) in the remaining places in the bounce integrals.
- [x] Partial summation (#1154, #1243) for the field line integral and chebyshev series.
- [x] Partial summation via FFT for the cubic spline.
- [x] Fold over field period for the cubic spline

### CODE REVIEW

File: desc/integrals/_bounce_utils.py [hunk 1]
```diff
@@ -779,26 +766,31 @@ def cubic_spline(T, f, Y, n0, n1, NFP=1, check=False):
...
+    num_transit = T.X
+    zeta = jnp.linspace(-1, 1, Y, endpoint=False)
+    # θ at uniformly spaced points in ζ along field line
+    theta = idct_non_uniform(zeta, T.cheb[..., jnp.newaxis, :], T.Y)
+    zeta = bijection_from_disc(zeta, T.domain[0], T.domain[1])
+
+    # f at uniformly spaced points in ζ along field line
+    f = _irfft2_non_uniform(
+        theta,
+        zeta,
```
Comment (unalmis) [inline desc/integrals/_bounce_utils.py:782]: Take 5 min tomorrow to use a uniform fft in zeta here so that we only need 1d non uniform transform

File: desc/integrals/_bounce_utils.py [hunk 2]
```diff
@@ -691,79 +679,93 @@ def fourier_chebyshev(theta, iota, alpha, num_transit):
...
+        θ over one toroidal transit.
     f : jnp.ndarray
+        Shape broadcasts with (num rho, 1, n_modes.size, m_modes.size).
         Fourier transform of f(θ, ζ) as returned by ``Bounce2D.fourier``.
-        ``n0=grid.num_theta``, ``n1=grid.num_zeta``, ``NFP=grid.NFP``.
     Y : int
-        Number of knots per transit to interpolate ``f``.
-    n0 : int
+        Number of knots per toroidal transit to interpolate ``f``.
+        This number will be rounded down to an integer multiple of NFP.
```
Comment (unalmis) [inline desc/integrals/_bounce_utils.py:762]: I confirmed this is the reason the baseline images for the neoclassical and fast ion confinement stuff changed. This can be confirmed easily by setting NFP to 1 everywhere in this function except for the line of code with `domain=(0, 2 * jnp.pi / NFP)`.

File: desc/integrals/bounce_integral.py
Comment (unalmis) [inline desc/integrals/bounce_integral.py:1]: None of these changes affect public API. Just internal stuff that was necessary to do the partial summation.

File: desc/integrals/_interp_utils.py [hunk 3]
```diff
@@ -137,24 +137,59 @@ def irfft_non_uniform(xq, a, n, domain=(0, 2 * jnp.pi), axis=-1):
...
+        if (n % 2) == 0:
+            i = (0, -1)
+        else:
+            i = 0
+        a = jnp.moveaxis(a, axis, -1).at[..., i].divide(2) * 2
+    vander = jnp.exp(1j * _modes * (xq - domain[0])[..., jnp.newaxis])
     return (vander * a).real.sum(axis=-1)
 
 
+def ifft_non_uniform(xq, a, domain=(0, 2 * jnp.pi), axis=-1, _modes=None):
```
[the remark this answers is not in the document]
Comment (unalmis) [reply desc/integrals/_interp_utils.py:162]: identical except the modes are different

File: desc/integrals/_bounce_utils.py [hunk 4]
```diff
@@ -779,26 +781,45 @@ def cubic_spline(T, f, Y, n0, n1, NFP=1, check=False):
...
+
+    # Let m, n denote the poloidal and toroidal Fourier resolution. We need to
+    # compute a set of 2D Fourier series each on uniform (non-uniform) in ζ (θ)
+    # tensor product grids of size
+    #   |𝛉|×|𝛇| where |𝛉| = num alpha × num transit × NFP and |𝛇| = Y/NFP.
+    # Partial summation via FFT is more efficient than direct evaluation when
+    # mn|𝛉||𝛇| > m log(|𝛇|) |𝛇| + m|𝛉||𝛇| or equivalently n|𝛉| > log|𝛇| + |𝛉|.
+
+    if num_zeta >= f.shape[-2] and num_zeta == f.shape[-2]:
+        # FIXME: This does not work unless num_zeta == f.shape[-2],
```
[the remark this answers is not in the document]
Comment (unalmis) [reply desc/integrals/_bounce_utils.py:801]: Yes. I did not figure it out.

File: desc/integrals/bounce_integral.py [hunk 5]
```diff
@@ -681,7 +691,36 @@ def integrate(
...
+    #     This is now possible.
+    #  4. Uiform FFT can be used in toroidal direction. Combined with partial
+    #     summation the interpolation becomes cheap.
+    #     (Same code as ``desc/integrals/_bounce_utils.py::cubic_spline``).
+    #  5. The quadrature points are no longer functions of the solutions
+    #     to the nonlinear equation λB = 1. In particular, all the ζ values
+    #     are constants throughout optimization. This makes AD cheaper.
+    #  6. Because the interpolation is now purely a function of ``num_transit``
+    #     and θ, rather than the apriori unknown number of bounce points, the
+    #     expensive JAX limitation in GitHub issue #1303 is avoided.
```
Comment (unalmis) [inline desc/integrals/bounce_integral.py:717]: @f0uriest thoughts?

File: desc/integrals/_bounce_utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (f0uriest) [reply desc/integrals/_bounce_utils.py:801]: Can you make an issue and add the number to the comment so we can track it

[the remark this answers is not in the document]
Comment (unalmis) [reply desc/integrals/_bounce_utils.py:801]: done

File: desc/integrals/bounce_integral.py (hunk 5 above)
Comment (f0uriest) [reply desc/integrals/bounce_integral.py:718]: It can work well as long as you can find a good function to subtract. The usual issue is that "easy" functions to subtract aren't particularly smooth so they slow convergence of the now nonsingular portion of the integral.

I think knosos does something similar to what you're describing by analytically integrating some approximation of the singularity at the bounce point? But they then still need to do adaptive quadrature to compute the non-singular part.

**Review by f0uriest — APPROVED** [review]

File: desc/integrals/bounce_integral.py (hunk 5 above)
Comment (unalmis) [reply desc/integrals/bounce_integral.py:718]: To clarify, for strongly singular integrals, I think fast convergence requires a change of variable like we do. Looking at the convergence plots in the bounce averaging paper, I can conclude that the above would not work well unless the singularity can be subtracted to high order accuracy and that subtracted quantity can be analytically integrated somehow. (So I do not intend for this alternative to be used for `Gamma_c` etc.).

For weakly singular integrals, however, (e.g. effective ripple, and the invariant J), suppose we subtracted the singularity with just first order accuracy (same taylor expansion in comment above). Because the original singularity was weak, some math shows that the resulting integrand is smooth (bounded derivative). Also, for symmetric wells, the resulting integrand is periodic. Of course, the derivative of the resulting integrand is **not** smooth.

Looking at the convergence plots in the bounce averaging paper for weakly singular integrals, I think that the penalty from applying a uniform quadrature to an integrand that is smooth (but whose derivative is not smooth) is probably less than ``num_pitch``>=50, so it could work better for the reasons discussed in the comments.

**Comment by unalmis** [comment]
Could someone override codecov to merge? thanks

**Review by rahulgaur104 — APPROVED** [review]

**Comment by rahulgaur104** [comment]
After the benchmarks run successfully.

**Comment by unalmis** [comment]
and make sure to click the down arrow adjacent to the merge button and select squash & merge please
